### PR TITLE
Fix namespace sync with missing avatar_sha256

### PR DESCRIPTION
**Pull request by mdellweg**
[noissue]

### CODE REVIEW

File: pulp_ansible/app/models.py [hunk 1]
```diff
@@ -332,15 +332,9 @@ class AnsibleNamespaceMetadata(Content):
...
-        if self.avatar_sha256:
-            avatar = self._artifacts.filter(sha256=self.avatar_sha256).first()
-            if avatar is None:
-                log.debug(
-                    f"Artifact({self.avatar_sha256}) is missing for namespace avatar "
-                    f"{self.name}:{self.metadata_sha256}"
-                )
-            return avatar
-
+        ca = self.contentartifact_set.filter(relative_path=f"{self.name}-avatar").first()
```
Comment (gerrod3) [inline pulp_ansible/app/models.py:335]: Can you check if we need a `select_related('artifact')`  here?

File: pulp_ansible/app/models.py [hunk 2]
```diff
@@ -332,16 +332,9 @@ class AnsibleNamespaceMetadata(Content):
...
-                log.debug(
-                    f"Artifact({self.avatar_sha256}) is missing for namespace avatar "
-                    f"{self.name}:{self.metadata_sha256}"
-                )
-            return avatar
-
-        return None
+        return self._artifacts.filter(
+            content_memberships__relative_path="hbvgfzyzud-avatar"
+        ).first()
```
Comment (mdellweg) [inline pulp_ansible/app/models.py:337]: I believe this actually produces the proper inner join right away.

```sql
SELECT "core_artifact"."pulp_id", "core_artifact"."pulp_created", "core_artifact"."pulp_last_updated", "core_artifact"."file", "core_artifact"."size", "core_artifact"."md5", "core_artifact"."sha1", "core_artifact"."sha224", "core_artifact"."sha256", "core_artifact"."sha384", "core_artifact"."sha512", "core_artifact"."timestamp_of_interest", "core_artifact"."pulp_domain_id"
FROM "core_artifact"
INNER JOIN "core_contentartifact" ON ("core_artifact"."pulp_id" = "core_contentartifact"."artifact_id")
WHERE ("core_contentartifact"."content_id" = %s AND "core_contentartifact"."relative_path" = %s)
```

File: pulp_ansible/app/models.py (hunk 1 above)
Comment (mdellweg) [reply pulp_ansible/app/models.py]: I updated the query to be about artifacts without indirection.

File: pulp_ansible/app/tasks/collections.py [hunk 3]
```diff
@@ -1154,13 +1154,15 @@ def _pre_save(self, batch):
...
-                        d_content.content.avatar_sha256 = da.artifact.sha256
+                    if False:
+                        # This code invalidates the digest.
+                        # Check to see if avatar failed to download, update metadata if so
+                        if da.deferred_download:
+                            d_content.d_artifacts = None
+                            d_content.content.avatar_sha256 = None
+                        # Check to see if upstream didn't have avatar_sha256 set
+                        elif d_content.content.avatar_sha256 is None:
+                            d_content.content.avatar_sha256 = da.artifact.sha256
```
Comment (gerrod3) [inline pulp_ansible/app/tasks/collections.py:1165]:
```suggestion
                      # Check to see if avatar failed to download, update metadata if so
                      if da.deferred_download:
                          d_content.d_artifacts = None
                          d_content.content.avatar_sha256 = None
                          d_content.content.metadata_sha256 = None
```
How about this? If `da.deferred_download` is true that means the avatar failed to download. We null the `avatar_sha256` and `metadata_sha256` (<-this gets recalculated on save), that way on next sync we will try download the image again.

Comment (mdellweg) [reply pulp_ansible/app/tasks/collections.py:1165]: Interesting idea. Let's try.

**Review by gerrod3 — APPROVED** [review]

File: functest_requirements.txt [hunk 4]
```diff
@@ -7,3 +7,4 @@ numpy
 requests
 proxy.py~=2.4.3
 trustme~=1.1.0
+pulp-smash @ git+https://github.com/pulp/pulp-smash.git
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply functest_requirements.txt:10]: Good question. It was some import. I should have checked whether it came from another plugins pytestplugin.
(This is all a mess.)

[the remark this answers is not in the document]
Comment (mdellweg) [reply functest_requirements.txt:10]: pulp_ansible/tests/functional/utils.py is smash infested...

**Pull request merged**